test/s3-tests/src/it/java/software/amazon/awssdk/services/s3/regression/UploadStreamingRegressionTesting.java

```diff
@@ -0,0 +1,753 @@
+import software.amazon.awssdk.transfer.s3.model.Upload;
+import software.amazon.awssdk.utils.CancellableOutputStream;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+import software.amazon.awssdk.utils.FunctionalUtils;
+import software.amazon.awssdk.utils.Logger;
+
+public class UploadStreamingRegressionTesting extends BaseS3RegressionTest {
+    private static final Logger LOG = Logger.loggerFor(UploadStreamingRegressionTesting.class);
+
+    private static final ExecutorService ASYNC_REQUEST_BODY_EXECUTOR = Executors.newSingleThreadExecutor();
+
+    static final byte[] smallContent = "Hello world".getBytes(StandardCharsets.UTF_8);
+    static final byte[] largeContent = largeContent();
+
+    private static Path testFileSmall;
+    private static Path testFileLarge;
+
+    @BeforeAll
+    static void setupClass() throws InterruptedException, IOException {
+        // Log.initLoggingToStdout(Log.LogLevel.Trace);
+
+        testFileSmall = S3ChecksumsTestUtils.createRandomFile16KB();
+        testFileLarge = S3ChecksumsTestUtils.createRandomFile80MB();
+    }
+
+    @AfterAll
+    public static void cleanup() {
+        ASYNC_REQUEST_BODY_EXECUTOR.shutdownNow();
+    }
+
+    @ParameterizedTest
+    @MethodSource("uploadConfigs")
+    void putObject(UploadConfig config) throws Exception {
+        assumeNotAccelerateWithPathStyle(config.getBaseConfig());
+        assumeNotAccessPointWithPathStyle(config.getBaseConfig());
+        assumeNotAccelerateWithArnType(config.getBaseConfig());
+        assumeNotAccelerateWithEoz(config.getBaseConfig());
+
+        // For testing purposes, ContentProvider is Publisher<ByteBuffer> for async clients
+        // There is no way to create AsyncRequestBody with a Publisher<ByteBuffer> and also provide the content length
+        Assumptions.assumeFalse(config.getBodyType() == BodyType.CONTENT_PROVIDER_WITH_LENGTH
+                                && config.getBaseConfig().getFlavor().isAsync(),
+                                "No way to create AsyncRequestBody by giving both an Publisher and the content length");
+
+        // Payload signing doesn't work correctly for async java based
+        Assumptions.assumeFalse(
+            (config.getBaseConfig().getFlavor() == S3ClientFlavor.ASYNC_JAVA_BASED ||
+             config.getBaseConfig().getFlavor() == S3ClientFlavor.TM_JAVA)
+            && (config.payloadSigning()
+                // MRAP requires body signing
+                || config.getBaseConfig().getBucketType() == BucketType.MRAP),
+            "Async payload signing doesn't work with Java based clients");
+
+        // For testing purposes, ContentProvider is Publisher<ByteBuffer> for async clients
+        // Async java based clients don't currently support unknown content-length bodies
+        Assumptions.assumeFalse(
+            (config.getBaseConfig().getFlavor() == S3ClientFlavor.ASYNC_JAVA_BASED ||
+             config.getBaseConfig().getFlavor() == S3ClientFlavor.TM_JAVA)
```

Comment:
This should be supported in multipart client, right?

Reply:
Do we have a test case covering multipart client + unknown content-length?

Reply:
Yes, we have `INPUTSTREAM_NO_LENGTH` which uses `AsyncRequestBody
                    .fromInputStream(AsyncRequestBodyFromInputStreamConfiguration)` without length and `CONTENT_PROVIDER_NO_LENGTH` which uses `AsyncRequestBody.fromPublisher(Publisher)` which are used with both multipart and single part